### PR TITLE
feat(autonat): Implement AutoNATv2

### LINKED ISSUE #4524: AutoNAT v2
Preliminary tracking issue around all the work that needs to happen around AutoNATv2 or issues that are associated with it.

```[tasklist]
### Implementation tasks
- [ ] https://github.com/libp2p/rust-libp2p/issues/4226
```
```[tasklist]
### Follow-up work
- [ ] https://github.com/libp2p/rust-libp2p/issues/3953
```
```[tasklist]
### Known bugs
- [ ] https://github.com/libp2p/rust-libp2p/issues/3986
- [ ] https://github.com/libp2p/rust-libp2p/issues/3889
- [ ] https://github.com/libp2p/rust-libp2p/issues/3308
- [ ] https://github.com/libp2p/rust-libp2p/issues/4873
- [ ] https://github.com/libp2p/rust-libp2p/issues/5059
```

### PR DESCRIPTION
**Pull request by umgefahren**
## Description

Closes: #4524

This is the implementation of the evolved AutoNAT protocol, named AutonatV2 as defined in the [spec](https://github.com/libp2p/specs/blob/03718ef0f2dea4a756a85ba716ee33f97e4a6d6c/autonat/autonat-v2.md).
The stabilization PR for the spec can be found under libp2p/specs#538.

The work on the Rust implementation can be found in the PR to my fork: umgefahren/rust-libp2p#1.

The implementation has been smoke-tested with the Go implementation (PR: libp2p/go-libp2p#2469).

The new protocol addresses shortcomings of the original AutoNAT protocol:

- Since the server now always dials back over a newly allocated port, this made #4568 necessary; the client can be sure of the reachability state for other peers, even if the connection to the server was made through a hole punch.
- The server can now test addresses different from the observed address (i.e., the connection to the server was made through a `p2p-circuit`). To mitigate against DDoS attacks, the client has to send more data to the server than the dial-back costs.

#### IPv4 & IPv6

There are open questions on how to handle cases where a client is only reachable over IPv4 or IPv6, but the server is connected to a network with an incompatible IP family.

Relevant comments in the other PR:

- [Original assessment](https://github.com/umgefahren/rust-libp2p/pull/1#issuecomment-2167215625)
- [@thomaseizinger assessment](https://github.com/umgefahren/rust-libp2p/pull/1#issuecomment-2168171137)
- [Decision](https://github.com/umgefahren/rust-libp2p/pull/1#issuecomment-2171116249)

The decision was made to ship the improvement in a later version, when we know more about the issue. It's still an improvement over AutoNATv1.

Handling these cases would be easier once rust-lang/rust#86442 is merged.

#### Picking algorithm

The spec proposes a picking algorithm for the address tested by the server. At the moment, the server is just using the first address sent by the client. An improvement to this algorithm can be implemented later.

### Open questions

- [x] Should we replace the deprecated v1 version in libp2p-server?

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

This work was sponsored by the Filecoin grant program: filecoin-project/devgrants#1421

### CODE REVIEW

File: examples/autonat/src/bin/autonat_client.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 // DEALINGS IN THE SOFTWARE.
 
 #![doc = include_str!("../../README.md")]
+#![allow(deprecated)]
```
Comment (jxs) [inline examples/autonat/src/bin/autonat_client.rs]: why do we need these `#![allow(deprecated)]`?

File: examples/autonatv2/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (jxs) [inline examples/autonatv2/src/lib.rs]: do we need this file?

**Comment by umgefahren** [comment]
@jxs this already got an approving review over on umgefahren/rust-libp2p#1 but it wouldn't hurt if you took a look at it before merging.

File: examples/autonatv2/src/lib.rs
Comment (thomaseizinger) [inline examples/autonatv2/src/lib.rs:1]: What is this file good for?

File: protocols/autonat/CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,13 @@
 - Due to the refactor of `Transport` it's no longer required to create a seperate transport for
 AutoNAT where port reuse is disabled. This information is now passed by the behaviour.
   See [PR 4568](https://github.com/libp2p/rust-libp2p/pull/4568).
+- Introduce the new AutoNATv2 protocol.
+  It's split into a client and a server part, represented in their respective modules
+  Features:
+    - Since the server now always dials back over a newly allocated port, this made refactor(*): Transport redesign #4568 necessary; the client can be sure of the reachability state for other peers, even if the connection to the server was made through a hole punch.
```
Comment (thomaseizinger) [inline protocols/autonat/CHANGELOG.md]:
```suggestion
    - The server now always dials back over a newly allocated port.
      This more accurately reflects the reachability state for other peers and avoids accidental hole punching.
```

File: protocols/autonat/CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,13 @@
...
+    - The server can now test addresses different from the observed address (i.e., the connection to the server was made through a `p2p-circuit`). To mitigate against DDoS attacks, the client has to send more data to the server than the dial-back costs.
+  See [PR 5526].
+[PR 5526]: https://github.com/libp2p/rust-libp2p/pull/5526
```
Comment (thomaseizinger) [inline protocols/autonat/CHANGELOG.md]:
```suggestion
  See [PR 5526](https://github.com/libp2p/rust-libp2p/pull/5526).

```

**Review by thomaseizinger — APPROVED** [review]
Nice work @umgefahren ! This has been vetted extensively in your fork already, so good to go from my end.

File: examples/autonatv2/src/lib.rs
Comment (umgefahren) [reply examples/autonatv2/src/lib.rs:1]: It's useless

**Review by jxs — APPROVED** [review]
Thanks for all the hard work Hannes, and again Thomas for the review! :heart:

**Comment by umgefahren** [comment]
> Thanks for all the hard work Hannes, and again Thomas for the review! ❤️

It was a pleasure and I'm so grateful we are here.

**Comment by b-zee** [comment]
I've been testing out this v2 protocol. When testing locally I see that my loopback address is being confirmed as external address. I'm curious if this is designed on purpose like that. E.g. would I need to remove the external address in logic for `ExternalAddrConfirmed`? Or is there a way to prevent the loopback address to be confirmed/probed at all?

I'd happily make an issue for this, but perhaps I am using the behaviour wrong as there isn't a lot of surface documentation on the usage. (The v1 protocol has built-in confidence logic, which I reckon we're supposed to implement ourselves in v2, which I fully stand behind, but just wondering how this is supposed to work.)

**Pull request merged**